lucet-concurrency-tests/src/killswitch.rs

```diff
@@ -0,0 +1,806 @@
+};
+use std::sync::Arc;
+
+use lucet_module::FunctionPointer;
+use lucet_runtime_internals::lock_testpoints::Syncpoint;
+use lucet_runtime_internals::module::Module;
+use lucet_runtime_internals::module::{MockExportBuilder, MockModuleBuilder};
+use lucet_runtime_internals::vmctx::lucet_vmctx;
+
+pub static mut ENTERING_GUEST: Option<Syncpoint> = None;
+
+/// A convenience wrapper around running our mock timeout module's `onetwothree` function,
+/// and asserting that it returned the expected result.
+pub fn run_onetwothree(inst: &mut Instance) {
+    let retval = inst
+        .run("onetwothree", &[])
+        .expect("instance runs")
+        .unwrap_returned();
+    assert_eq!(libc::c_int::from(retval), 123);
+}
+
+/// Construct a Lucet module with surface area for testing all kinds of KillSwitch termination
+/// cases.
+///
+/// This includes:
+/// * Normal guest execution
+/// * Hostcall execution
+/// * Yield/resume
+/// * Guest fault
+/// * Hostcall fault
+pub fn mock_killswitch_module() -> Arc<dyn Module> {
+    extern "C" fn onetwothree(_vmctx: *mut lucet_vmctx) -> std::os::raw::c_int {
+        123
+    }
+
+    extern "C" fn run_guest(_vmctx: *mut lucet_vmctx) {
+        unsafe {
+            // ENTERING_GUEST is only populated if the test requires this syncpoint be checked.
+            if let Some(entering_guest) = ENTERING_GUEST.as_ref() {
+                entering_guest.check();
+            }
+        }
+    }
+
+    extern "C" fn infinite_loop(_vmctx: *mut lucet_vmctx) {
+        unsafe {
+            // ENTERING_GUEST is only populated if the test requires this syncpoint be checked.
+            if let Some(entering_guest) = ENTERING_GUEST.as_ref() {
+                entering_guest.check();
+            }
+        }
+        loop {}
+    }
+
+    extern "C" fn fatal(vmctx: *mut lucet_vmctx) {
+        extern "C" {
+            fn lucet_vmctx_get_heap(vmctx: *mut lucet_vmctx) -> *mut u8;
+        }
+
+        unsafe {
+            let heap_base = lucet_vmctx_get_heap(vmctx);
+
+            // Using the default limits, each instance as of this writing takes up 0x200026000 bytes
+            // worth of virtual address space. We want to access a point beyond all the instances,
+            // so that memory is unmapped. We assume no more than 16 instances are mapped
+            // concurrently. This may change as the library, test configuration, linker, phase of
+            // moon, etc change, but for now it works.
+            *heap_base.offset(0x0002_0002_6000 * 16) = 0;
+        }
+    }
+
+    extern "C" fn hit_sigstack_guard_page(vmctx: *mut lucet_vmctx) {
+        extern "C" {
+            fn lucet_vmctx_get_globals(vmctx: *mut lucet_vmctx) -> *mut u8;
+        }
+
+        unsafe {
+            let globals_base = lucet_vmctx_get_globals(vmctx);
+
+            // Using the default limits, the globals are a page; try to write just off the end
+            *globals_base.offset(0x1000) = 0;
```

Comment:
I wish I had mentioned this before, but I missed it.  I do apologize.

I'm curious about where this default 0x1000 lives.  I'm not seeing it over at: https://github.com/bytecodealliance/lucet/blob/83fd2248b36c780cc74c0abb49024cca4a0c053c/lucet-runtime/lucet-runtime-internals/src/alloc/mod.rs#L461

Reply:
Also.  Hex is not decimal.  Yep.  Thanks peer shame.

Reply:
https://www.youtube.com/watch?v=OjYoNL4g5Vg